### DISCUSSION TITLE
Course to Batch to Learner flow

### DISCUSSION THREAD

**Discussion by vishaldeltech** [Sunbird-Ed/Community#410] [General Discussion]
Hi folks

We have a business case of custom batch-level assets added to a course (visible only to that specific batch), and learner-level assets visible only to a specific learner. Due dates, and class timings can also be set up at batch/learner level relative to an event. We still need course updates to flow to the learner. We are thinking of meeting the business requirement by managing 3 layers for course consumption. Here are the details of the 3 layers

#### Layer 1 at Course
In the existing course structure, we add the following capabilities

- Sequential unlock, and condition-based unlock. This works for both nodes and content.
- Due dates/unlock date configuration relative to a date variable (for example, batch start date, learner joining date, unlock/completion date of specific node/content)

#### Layer 2 at Batch
Batch maintains a separate layer that saves modifications for each node/content of the course. The modifications can allow the following to be controlled at the batch level

- Unlocking any specific asset
- Calculate and set actual dates at any node/content that required a batch-level variable (for example, batch start date)
- Manually set a specific date for a node/content 
- Custom nodes/assets for the batch at a specific node of the course

#### Layer 3 at Learner
At the learner level, another layer is maintained that saves modifications at the learner level. These modifications apply over the layer maintained by the batch

- Unlocking any specific asset
- Evaluate actual dates at any node/content that required a learner-level variable (for example, learner joining date)
- Manually set a specific date for a content 
- Custom nodes/assets for the learner at a specific node of the course
- Progress on content 

### How the 3 layers work
At runtime when a learner accesses a course, 

1. the system picks the course layer
2. It applies batch layer configuration over course layer
3. It applies learner layer over result of step 2
4. Sends the resulting JSON to the front end

### What happens if the course gets updated after publish
The batch and learner layer work with node/content references in the course. So any modification moves with changes in the structure of the course. For unresolved conflicts, if any, the batch manager will have the fix them for their batch

Requesting feedback from the community on the approach

**Comment by Krishnaj20** [comment]
@vishaldeltech - Course unlock should be at a batch level instead of course level. For example, A course can have multiple batches and each batch can have a different unlock schedule.  It will be useful if there are any visual flows depicting how the functional usecases will be enabled by this architecture. 

@rhwarrier @reshmi-nair @vrayulu @maheshkumargangula @pallakartheekreddy - Please go thru this proposal and share your feedback.

**Comment by rhwarrier** [comment]
The whole idea of bring in the batch construct is to be able to mangage specifics for the cohort - including start/ end dates, rules of completion, choice of content served via the course,  etc. Maybe you could look at adding some more rules based on user role/ sub-role etc. But you should also go back at take a look at the data  - especially sub-roles. We need to ensure that the data that is available is clean and usable. (it is a randomly growing field, from what I remember)  

Going down to learner level assets sounds like a bit of an overkill - if there are a group of similar learners who need access to a specific set of assets, won't it make sense to have a batch for such a group ? Learner level configurability points to a need for personalised learning journeys. Maybe we should not try to force fit the existing course construct for such a use case.  
I'm sure the DC will have more specifics to add.

**Comment by Krishnaj20** [comment]
@parthlawate @rhwarrier @reshmi-nair @vishaldeltech @mathewjpallan @maheshkumargangula - Sharing the confluence docs as discussed in today's meeting:
Use cases: https://project-sunbird.atlassian.net/wiki/spaces/SUN/pages/3296034912/Course+Consumption+Modalities
Tech Design: https://project-sunbird.atlassian.net/wiki/spaces/SUN/pages/3296264262/Tech+Design

@vrayulu

**Reply by vishaldeltech** [reply]
HI @Krishnaj20. I am seeing the message "You’ve stumbled on a restricted space". Please advise on how to proceed

![image](https://user-images.githubusercontent.com/67492325/224637016-c1e646ab-eaf5-4d02-be57-07156b3a4c62.png)

**Reply by Krishnaj20** [reply]
@vishaldeltech - Pls use these links:
Use cases: https://project-sunbird.atlassian.net/wiki/spaces/SUN/pages/3296034912/Course+Consumption+Modalities
Tech Design: https://project-sunbird.atlassian.net/wiki/spaces/SUN/pages/3296264262/Tech+Design

Let me know if you face any issues.

**Reply by vishaldeltech** [reply]
@Krishnaj20 I do not have access to this page. I am seeing the restricted space issue. If I create a page, it redirects to CQUB space

**Reply by mathewjpallan** [reply]
@vishaldeltech pls login with any gmail id/signup and you should be able to access

**Reply by Krishnaj20** [reply]
@vishaldeltech - are you able to access the confluence links?
